src/search/search_session.cpp

```diff
@@ -0,0 +1,869 @@
+﻿/*
+===========================================================================
+
+  Copyright (c) 2023 LandSandBoat Dev Teams
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see http://www.gnu.org/licenses/
+
+===========================================================================
+*/
+
+#include "search_session.h"
+#include "common/md52.h"
+#include "common/socket.h" // for ref<T>
+#include "common/utils.h"  // for unpack/pack bits
+#include "data_loader.h"
+#include <unordered_set>
+
+#include "packets/auction_history.h"
+#include "packets/auction_list.h"
+#include "packets/linkshell_list.h"
+#include "packets/party_list.h"
+#include "packets/search_comment.h"
+#include "packets/search_list.h"
+
+
+// A single IP should only have one request in flight at a time, so we are going to
+// be tracking the IP addresses of incoming requests and if we haven't cleared the
+// record for it - we block until its done
+extern shared_guarded<std::unordered_set<std::string>> gIPAddressesInUse;
```

Comment:
Instead of globals and externs, can we make gIPAddressesInUse a member of search_server and pass references to it around? Acess to this is shared_guarded so it's safe to access from multiple threads